File: web-client/src/services/tablePartCommandManager.ts
/**
 * Table Part Command Manager Service for Web Client.
 * 
 * This module provides command management functionality for table parts,
 * bridging panel buttons with form methods and handling command discovery,
 * registration, and execution.
 * 
 * Requirements: 2.1, 2.2, 2.3, 2.4, 2.5
 */

// ============================================================================
// Types and Interfaces
// ============================================================================

export enum CommandAvailability {
  ALWAYS = 'always',
  REQUIRES_SELECTION = 'requires_selection',
  REQUIRES_ROWS = 'requires_rows',
  REQUIRES_EDIT_MODE = 'requires_edit_mode',
  CUSTOM = 'custom'
}

export interface FormCommand {
  id: string
  name: string
  methodName: string
  formInstance: any
  availability: CommandAvailability
  parameters: any[]
  enabled: boolean
  customAvailabilityCheck?: () => boolean
}

export interface CommandContext {
  selectedRows: number[]
  tableData: any[]
  currentRow?: number
  currentColumn?: string
  additionalData: Record<string, any>
}

export interface CommandResult {
  success: boolean
  message?: string
  data?: any
  affectedRows: number[]
  refreshRequired: boolean
}

export interface ICommandDiscovery {
  discoverCommands(formInstance: any): FormCommand[]
}

// ============================================================================
// Command Discovery Strategies
// ============================================================================

export class AttributeBasedDiscovery implements ICommandDiscovery {
  /**
   * Discovers commands by looking for methods with tableCommand metadata.
   * 
   * Methods should be decorated with @tableCommand decorator.
   */
  discoverCommands(formInstance: any): FormCommand[] {
    const commands: FormCommand[] = []
    
    // Check if form instance has command metadata
    if (formInstance._tableCommands) {
      for (const [methodName, config] of Object.entries(formInstance._tableCommands)) {
        const command: FormCommand = {
          id: (config as any).id || methodName,
          name: (config as any).name || this.formatMethodName(methodName),
          methodName,
          formInstance,
          availability: (config as any).availability || CommandAvailability.ALWAYS,
          parameters: (config as any).parameters || [],
          enabled: (config as any).enabled !== false,
          customAvailabilityCheck: (config as any).customAvailabilityCheck
        }
        commands.push(command)
      }
    }
    
    // Also check the prototype for decorated methods
    const prototype = Object.getPrototypeOf(formInstance)
    if (prototype && prototype._tableCommands) {
      for (const [methodName, config] of Object.entries(prototype._tableCommands)) {
        // Skip if already found in instance
        if (formInstance._tableCommands && formInstance._tableCommands[methodName]) {
          continue
        }
        
        const command: FormCommand = {
          id: (config as any).id || methodName,
          name: (config as any).name || this.formatMethodName(methodName),
          methodName,
          formInstance,
          availability: (config as any).availability || CommandAvailability.ALWAYS,
          parameters: (config as any).parameters || [],
          enabled: (config as any).enabled !== false,
          customAvailabilityCheck: (config as any).customAvailabilityCheck
        }
        commands.push(command)
      }
    }
    
    return commands
  }
  
  private formatMethodName(methodName: string): string {
    return methodName.replace(/([A-Z])/g, ' $1').replace(/^./, str => str.toUpperCase())
  }
}

export class NamingConventionDiscovery implements ICommandDiscovery {
  /**
   * Discovers commands based on method naming conventions.
   */
  
  private readonly COMMAND_PATTERNS: Record<string, string[]> = {
    add_row: ['addRow', 'addTableRow', 'insertRow', 'createRow'],
    delete_row: ['deleteRow', 'removeRow', 'deleteSelected', 'removeSelected'],
    move_up: ['moveUp', 'moveRowUp', 'rowUp'],
    move_down: ['moveDown', 'moveRowDown', 'rowDown'],
    import_data: ['importData', 'importRows', 'loadData'],
    export_data: ['exportData', 'exportRows', 'saveData'],
    print_data: ['printData', 'printTable', 'printRows']
  }
  
  discoverCommands(formInstance: any): FormCommand[] {
    const commands: FormCommand[] = []
    
    for (const [commandId, patterns] of Object.entries(this.COMMAND_PATTERNS)) {
      for (const pattern of patterns) {
        if (typeof formInstance[pattern] === 'function') {
          const command: FormCommand = {
            id: commandId,
            name: this.getDefaultName(commandId),
            methodName: pattern,
            formInstance,
            availability: this.getDefaultAvailability(commandId),
            parameters: [],
            enabled: true
          }
          commands.push(command)
          break // Use first matching pattern
        }
      }
    }
    
    return commands
  }
  
  private getDefaultAvailability(commandId: string): CommandAvailability {
    if (['delete_row', 'move_up', 'move_down'].includes(commandId)) {
      return CommandAvailability.REQUIRES_SELECTION
    } else if (['export_data'].includes(commandId)) {
      return CommandAvailability.REQUIRES_ROWS
    } else {
      return CommandAvailability.ALWAYS
    }
  }
  
  private getDefaultName(commandId: string): string {
    const names: Record<string, string> = {
      add_row: 'Добавить строку',
      delete_row: 'Удалить строки',
      move_up: 'Переместить выше',
      move_down: 'Переместить ниже',
      import_data: 'Импорт данных',
      export_data: 'Экспорт данных',
      print_data: 'Печать данных'
    }
    return names[commandId] || commandId.replace(/_/g, ' ')
  }
}

// ============================================================================
// Command Manager
// ============================================================================

export class TablePartCommandManager {
  /**
   * Command manager for table parts.
   * 
   * Provides functionality to:
   * - Discover commands from form instances
   * - Register and manage form commands
   * - Execute commands with proper context
   * - Handle command availability and state updates
   * 
   * Requirements: 2.1, 2.2, 2.3, 2.4, 2.5
   */
  
  private registeredCommands: Map<string, FormCommand> = new Map()
  private discoveryStrategies: ICommandDiscovery[]
  private commandStateCache: Map<string, boolean> = new Map()
  private lastContext?: CommandContext
  
  constructor(discoveryStrategies?: ICommandDiscovery[]) {
    this.discoveryStrategies = discoveryStrategies || [
      new AttributeBasedDiscovery(),
      new NamingConventionDiscovery()
    ]
  }
  
  /**
   * Discover and register commands from a form instance.
   * 
   * Requirements: 2.1, 2.2
   */
  discoverAndRegisterCommands(formInstance: any): FormCommand[] {
    const discoveredCommands: FormCommand[] = []
    
    for (const strategy of this.discoveryStrategies) {
      try {
        const commands = strategy.discoverCommands(formInstance)
        discoveredCommands.push(...commands)
        console.log(`Discovered ${commands.length} commands using ${strategy.constructor.name}`)
      } catch (error) {
        console.warn(`Command discovery failed with ${strategy.constructor.name}:`, error)
      }
    }
    
    // Register discovered commands
    for (const command of discoveredCommands) {
      this.registerCommand(command)
    }
    
    console.log(`Total registered commands: ${this.registeredCommands.size}`)
    return discoveredCommands
  }
  
  /**
   * Register a form command.
   * 
   * Requirements: 2.2
   */
  registerCommand(command: FormCommand): void {
    this.registeredCommands.set(command.id, command)
    console.debug(`Registered command: ${command.id} -> ${command.methodName}`)
  }
  
  /**
   * Unregister a form command.
   */
  unregisterCommand(commandId: string): void {
    if (this.registeredCommands.has(commandId)) {
      this.registeredCommands.delete(commandId)
      console.debug(`Unregistered command: ${commandId}`)
    }
  }
  
  /**
   * Execute a registered command.
   * 
   * Requirements: 2.3, 2.4
   */
  async executeCommand(commandId: string, context: CommandContext): Promise<CommandResult> {
    const command = this.registeredCommands.get(commandId)
    
    if (!command) {
      return {
        success: false,
        message: `Command '${commandId}' not found`,
        affectedRows: [],
        refreshRequired: false
      }
    }
    
    // Check availability
    if (!this.isCommandAvailable(command, context)) {
      return {
        success: false,
        message: `Command '${commandId}' is not available in current context`,
        affectedRows: [],
        refreshRequired: false
      }
    }
    
    try {
      // Get the method from form instance
      const method = command.formInstance[command.methodName]
      
      if (typeof method !== 'function') {
        throw new Error(`Method '${command.methodName}' is not a function`)
      }
      
      // Prepare method arguments
      const args = this.prepareMethodArguments(method, command, context)
      
      // Execute the method
      const result = await method.apply(command.formInstance, args)
      
      // Handle different return types
      if (this.isCommandResult(result)) {
        return result
      } else if (typeof result === 'boolean') {
        return {
          success: result,
          message: result ? 'Command executed' : 'Command failed',
          affectedRows: [],
          refreshRequired: false
        }
      } else {
        return {
          success: true,
          message: 'Command executed successfully',
          data: result,
          affectedRows: [],
          refreshRequired: false
        }
      }
    } catch (error) {
      console.error(`Command execution failed for '${commandId}':`, error)
      return {
        success: false,
        message: `Command execution failed: ${error instanceof Error ? error.message : String(error)}`,
        affectedRows: [],
        refreshRequired: false
      }
    }
  }
  
  private prepareMethodArguments(method: Function, command: FormCommand, context: CommandContext): any[] {
    const args: any[] = []
    
    // Add predefined parameters
    args.push(...command.parameters)
    
    // For JavaScript/TypeScript, we can't easily inspect method signatures
    // So we'll pass context as the first argument if no predefined parameters
    if (command.parameters.length === 0) {
      args.push(context)
    }
    
    return args
  }
  
  private isCommandResult(obj: any): obj is CommandResult {
    return obj && typeof obj === 'object' && 'success' in obj
  }
  
  /**
   * Update command availability states.
   * 
   * Requirements: 2.5
   */
  updateCommandStates(context: CommandContext): Record<string, boolean> {
    this.lastContext = context
    const states: Record<string, boolean> = {}
    
    for (const [commandId, command] of this.registeredCommands) {
      const isAvailable = this.isCommandAvailable(command, context)
      states[commandId] = isAvailable
      this.commandStateCache.set(commandId, isAvailable)
    }
    
    return states
  }
  
  private isCommandAvailable(command: FormCommand, context: CommandContext): boolean {
    if (!command.enabled) {
      return false
    }
    
    if (command.customAvailabilityCheck) {
      return command.customAvailabilityCheck()
    }
    
    switch (command.availability) {
      case CommandAvailability.ALWAYS:
        return true
      case CommandAvailability.REQUIRES_SELECTION:
        return context.selectedRows.length > 0
      case CommandAvailability.REQUIRES_ROWS:
        return context.tableData.length > 0
      case CommandAvailability.REQUIRES_EDIT_MODE:
        return context.additionalData.editMode === true
      default:
        return true
    }
  }
  
  /**
   * Get cached command availability state.
   */
  getCommandState(commandId: string): boolean {
    return this.commandStateCache.get(commandId) ?? false
  }
  
  /**
   * Get list of currently available commands.
   */
  getAvailableCommands(context?: CommandContext): FormCommand[] {
    const checkContext = context || this.lastContext
    if (!checkContext) {
      return Array.from(this.registeredCommands.values())
    }
    
    const available: FormCommand[] = []
    for (const command of this.registeredCommands.values()) {
      if (this.isCommandAvailable(command, checkContext)) {
        available.push(command)
      }
    }
    
    return available
  }
  
  /**
   * Get all registered commands.
   */
  getRegisteredCommands(): Record<string, FormCommand> {
    const result: Record<string, FormCommand> = {}
    for (const [id, command] of this.registeredCommands) {
      result[id] = command
    }
    return result
  }
  
  /**
   * Clear all registered commands.
   */
  clearCommands(): void {
    this.registeredCommands.clear()
    this.commandStateCache.clear()
    this.lastContext = undefined
  }
}

// ============================================================================
// Decorator for TypeScript/JavaScript
// ============================================================================

export interface TableCommandConfig {
  id?: string
  name?: string
  availability?: CommandAvailability
  parameters?: any[]
  enabled?: boolean
  customAvailabilityCheck?: () => boolean
}

/**
 * Decorator to mark methods as table commands.
 * 
 * Example:
 *   @tableCommand({ id: 'add_row', name: 'Add Row', availability: CommandAvailability.ALWAYS })
 *   addNewRow(context: CommandContext) {
 *     // Implementation
 *   }
 */
export function tableCommand(config: TableCommandConfig = {}) {
  return function (target: any, propertyKey: string | symbol, descriptor: PropertyDescriptor) {
    // Initialize _tableCommands if it doesn't exist on the prototype
    if (!target.constructor.prototype._tableCommands) {
      target.constructor.prototype._tableCommands = {}
    }
    
    // Convert propertyKey to string if it's a symbol
    const methodName = typeof propertyKey === 'string' ? propertyKey : propertyKey.toString()
    
    // Store command configuration on the prototype
    target.constructor.prototype._tableCommands[methodName] = {
      id: config.id || methodName,
      name: config.name || methodName.replace(/([A-Z])/g, ' $1').replace(/^./, str => str.toUpperCase()),
      availability: config.availability || CommandAvailability.ALWAYS,
      parameters: config.parameters || [],
      enabled: config.enabled !== false,
      customAvailabilityCheck: config.customAvailabilityCheck
    }
    
    return descriptor
  }
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Create a command manager with default discovery strategies.
 */
export function createCommandManager(): TablePartCommandManager {
  return new TablePartCommandManager()
}

/**
 * Create a command context from table part state.
 */
export function createCommandContext(
  selectedRows: number[] = [],
  tableData: any[] = [],
  additionalData: Record<string, any> = {}
): CommandContext {
  return {
    selectedRows,
    tableData,
    additionalData
  }
}

File: src/services/table_part_command_manager.py
"""
Table Part Command Manager Service.

This module provides command management functionality for table parts,
bridging panel buttons with form methods and handling command discovery,
registration, and execution.

Requirements: 2.1, 2.2, 2.3, 2.4, 2.5
"""

from typing import Dict, List, Optional, Callable, Any, Union
from dataclasses import dataclass, field
from abc import ABC, abstractmethod
import inspect
import logging
from enum import Enum

logger = logging.getLogger(__name__)


class CommandAvailability(Enum):
    """Command availability states"""
    ALWAYS = "always"
    REQUIRES_SELECTION = "requires_selection"
    REQUIRES_ROWS = "requires_rows"
    REQUIRES_EDIT_MODE = "requires_edit_mode"
    CUSTOM = "custom"


@dataclass
class FormCommand:
    """Configuration for a form command"""
    id: str
    name: str
    method_name: str
    form_instance: Any
    availability: CommandAvailability = CommandAvailability.ALWAYS
    parameters: List[Any] = field(default_factory=list)
    enabled: bool = True
    custom_availability_check: Optional[Callable[[], bool]] = None
    
    def is_available(self, context: 'CommandContext') -> bool:
        """Check if command is available in current context"""
        if not self.enabled:
            return False
        
        if self.custom_availability_check:
            return self.custom_availability_check()
        
        if self.availability == CommandAvailability.ALWAYS:
            return True
        elif self.availability == CommandAvailability.REQUIRES_SELECTION:
            return len(context.selected_rows) > 0
        elif self.availability == CommandAvailability.REQUIRES_ROWS:
            return len(context.table_data) > 0
        elif self.availability == CommandAvailability.REQUIRES_EDIT_MODE:
            return context.additional_data.get('edit_mode', False)
        
        return True


@dataclass
class CommandContext:
    """Context information for command execution"""
    selected_rows: List[int] = field(default_factory=list)
    table_data: List[Dict[str, Any]] = field(default_factory=list)
    current_row: Optional[int] = None
    current_column: Optional[str] = None
    additional_data: Dict[str, Any] = field(default_factory=dict)


@dataclass
class CommandResult:
    """Result of command execution"""
    success: bool
    message: Optional[str] = None
    data: Any = None
    affected_rows: List[int] = field(default_factory=list)
    refresh_required: bool = False


class ICommandDiscovery(ABC):
    """Interface for command discovery strategies"""
    
    @abstractmethod
    def discover_commands(self, form_instance: Any) -> List[FormCommand]:
        """Discover commands from a form instance"""
        pass


class AttributeBasedDiscovery(ICommandDiscovery):
    """Discovers commands based on method attributes"""
    
    def discover_commands(self, form_instance: Any) -> List[FormCommand]:
        """
        Discover commands by looking for methods with table_command attribute.
        
        Methods should be decorated with @table_command decorator.
        """
        commands = []
        
        for method_name in dir(form_instance):
            if method_name.startswith('_'):
                continue
                
            method = getattr(form_instance, method_name)
            if not callable(method):
                continue
            
            # Check if method has table_command attribute
            if hasattr(method, '_table_command_config'):
                config = method._table_command_config
                command = FormCommand(
                    id=config.get('id', method_name),
                    name=config.get('name', method_name.replace('_', ' ').title()),
                    method_name=method_name,
                    form_instance=form_instance,
                    availability=config.get('availability', CommandAvailability.ALWAYS),
                    parameters=config.get('parameters', []),
                    enabled=config.get('enabled', True)
                )
                commands.append(command)
        
        return commands


class NamingConventionDiscovery(ICommandDiscovery):
    """Discovers commands based on naming conventions"""
    
    COMMAND_PATTERNS = {
        'add_row': ('add_row', 'add_table_row', 'insert_row', 'create_row'),
        'delete_row': ('delete_row', 'remove_row', 'delete_selected', 'remove_selected'),
        'move_up': ('move_up', 'move_row_up', 'row_up'),
        'move_down': ('move_down', 'move_row_down', 'row_down'),
        'import_data': ('import_data', 'import_rows', 'load_data'),
        'export_data': ('export_data', 'export_rows', 'save_data'),
        'print_data': ('print_data', 'print_table', 'print_rows')
    }
    
    def discover_commands(self, form_instance: Any) -> List[FormCommand]:
        """Discover commands based on method naming patterns"""
        commands = []
        
        for command_id, patterns in self.COMMAND_PATTERNS.items():
            for pattern in patterns:
                if hasattr(form_instance, pattern):
                    method = getattr(form_instance, pattern)
                    if callable(method):
                        # Determine availability based on command type
                        availability = self._get_default_availability(command_id)
                        
                        command = FormCommand(
                            id=command_id,
                            name=self._get_default_name(command_id),
                            method_name=pattern,
                            form_instance=form_instance,
                            availability=availability
                        )
                        commands.append(command)
                        break  # Use first matching pattern
        
        return commands
    
    def _get_default_availability(self, command_id: str) -> CommandAvailability:
        """Get default availability for standard commands"""
        if command_id in ['delete_row', 'move_up', 'move_down']:
            return CommandAvailability.REQUIRES_SELECTION
        elif command_id in ['export_data']:
            return CommandAvailability.REQUIRES_ROWS
        else:
            return CommandAvailability.ALWAYS
    
    def _get_default_name(self, command_id: str) -> str:
        """Get default display name for standard commands"""
        names = {
            'add_row': 'Добавить строку',
            'delete_row': 'Удалить строки',
            'move_up': 'Переместить выше',
            'move_down': 'Переместить ниже',
            'import_data': 'Импорт данных',
            'export_data': 'Экспорт данных',
            'print_data': 'Печать данных'
        }
        return names.get(command_id, command_id.replace('_', ' ').title())


class TablePartCommandManager:
    """
    Command manager for table parts.
    
    Provides functionality to:
    - Discover commands from form instances
    - Register and manage form commands
    - Execute commands with proper context
    - Handle command availability and state updates
    
    Requirements: 2.1, 2.2, 2.3, 2.4, 2.5
    """
    
    def __init__(self, discovery_strategies: Optional[List[ICommandDiscovery]] = None):
        """
        Initialize command manager.
        
        Args:
            discovery_strategies: List of command discovery strategies to use
        """
        self.registered_commands: Dict[str, FormCommand] = {}
        self.discovery_strategies = discovery_strategies or [
            AttributeBasedDiscovery(),
            NamingConventionDiscovery()
        ]
        self.command_state_cache: Dict[str, bool] = {}
        self.last_context: Optional[CommandContext] = None
    
    def discover_and_register_commands(self, form_instance: Any) -> List[FormCommand]:
        """
        Discover and register commands from a form instance.
        
        Args:
            form_instance: Form instance to discover commands from
            
        Returns:
            List of discovered and registered commands
            
        Requirements: 2.1, 2.2
        """
        discovered_commands = []
        
        for strategy in self.discovery_strategies:
            try:
                commands = strategy.discover_commands(form_instance)
                discovered_commands.extend(commands)
                logger.info(f"Discovered {len(commands)} commands using {strategy.__class__.__name__}")
            except Exception as e:
                logger.warning(f"Command discovery failed with {strategy.__class__.__name__}: {e}")
        
        # Register discovered commands
        for command in discovered_commands:
            self.register_command(command)
        
        logger.info(f"Total registered commands: {len(self.registered_commands)}")
        return discovered_commands
    
    def register_command(self, command: FormCommand):
        """
        Register a form command.
        
        Args:
            command: Form command to register
            
        Requirements: 2.2
        """
        self.registered_commands[command.id] = command
        logger.debug(f"Registered command: {command.id} -> {command.method_name}")
    
    def unregister_command(self, command_id: str):
        """
        Unregister a form command.
        
        Args:
            command_id: ID of command to unregister
        """
        if command_id in self.registered_commands:
            del self.registered_commands[command_id]
            logger.debug(f"Unregistered command: {command_id}")
    
    def execute_command(self, command_id: str, context: CommandContext) -> CommandResult:
        """
        Execute a registered command.
        
        Args:
            command_id: ID of command to execute
            context: Execution context
            
        Returns:
            Command execution result
            
        Requirements: 2.3, 2.4
        """
        if command_id not in self.registered_commands:
            return CommandResult(
                success=False,
                message=f"Command '{command_id}' not found"
            )
        
        command = self.registered_commands[command_id]
        
        # Check availability
        if not command.is_available(context):
            return CommandResult(
                success=False,
                message=f"Command '{command_id}' is not available in current context"
            )
        
        try:
            # Get the method from form instance
            method = getattr(command.form_instance, command.method_name)
            
            # Prepare method arguments
            args = self._prepare_method_arguments(method, command, context)
            
            # Execute the method
            result = method(*args)
            
            # Handle different return types
            if isinstance(result, CommandResult):
                return result
            elif isinstance(result, bool):
                return CommandResult(
                    success=result,
                    message="Command executed" if result else "Command failed"
                )
            else:
                return CommandResult(
                    success=True,
                    message="Command executed successfully",
                    data=result
                )
        
        except Exception as e:
            logger.error(f"Command execution failed for '{command_id}': {e}")
            return CommandResult(
                success=False,
                message=f"Command execution failed: {str(e)}"
            )
    
    def _prepare_method_arguments(
        self,
        method: Callable,
        command: FormCommand,
        context: CommandContext
    ) -> List[Any]:
        """Prepare arguments for method execution based on signature"""
        sig = inspect.signature(method)
        args = []
        
        # Add predefined parameters
        args.extend(command.parameters)
        
        # Add context-based parameters if method expects them
        for param_name, param in sig.parameters.items():
            if param_name == 'self':
                continue
            elif param_name == 'context':
                args.append(context)
            elif param_name == 'selected_rows':
                args.append(context.selected_rows)
            elif param_name == 'table_data':
                args.append(context.table_data)
            elif param_name == 'current_row':
                args.append(context.current_row)
            elif param_name == 'current_column':
                args.append(context.current_column)
        
        return args
    
    def update_command_states(self, context: CommandContext) -> Dict[str, bool]:
        """
        Update command availability states.
        
        Args:
            context: Current context for availability checking
            
        Returns:
            Dictionary mapping command IDs to availability states
            
        Requirements: 2.5
        """
        self.last_context = context
        states = {}
        
        for command_id, command in self.registered_commands.items():
            is_available = command.is_available(context)
            states[command_id] = is_available
            self.command_state_cache[command_id] = is_available
        
        return states
    
    def get_command_state(self, command_id: str) -> bool:
        """
        Get cached command availability state.
        
        Args:
            command_id: ID of command to check
            
        Returns:
            True if command is available, False otherwise
        """
        return self.command_state_cache.get(command_id, False)
    
    def get_available_commands(self, context: Optional[CommandContext] = None) -> List[FormCommand]:
        """
        Get list of currently available commands.
        
        Args:
            context: Context to check availability against. Uses last context if None.
            
        Returns:
            List of available commands
        """
        check_context = context or self.last_context
        if not check_context:
            return list(self.registered_commands.values())
        
        available = []
        for command in self.registered_commands.values():
            if command.is_available(check_context):
                available.append(command)
        
        return available
    
    def get_registered_commands(self) -> Dict[str, FormCommand]:
        """Get all registered commands"""
        return self.registered_commands.copy()
    
    def clear_commands(self):
        """Clear all registered commands"""
        self.registered_commands.clear()
        self.command_state_cache.clear()
        self.last_context = None


# Decorator for marking methods as table commands
def table_command(
    command_id: Optional[str] = None,
    name: Optional[str] = None,
    availability: CommandAvailability = CommandAvailability.ALWAYS,
    parameters: Optional[List[Any]] = None,
    enabled: bool = True
):
    """
    Decorator to mark methods as table commands.
    
    Args:
        command_id: Unique identifier for the command
        name: Display name for the command
        availability: When the command is available
        parameters: Default parameters for the command
        enabled: Whether the command is enabled by default
    
    Example:
        @table_command(command_id='add_row', name='Add Row', availability=CommandAvailability.ALWAYS)
        def add_new_row(self, context: CommandContext):
            # Implementation
            pass
    """
    def decorator(func):
        func._table_command_config = {
            'id': command_id or func.__name__,
            'name': name or func.__name__.replace('_', ' ').title(),
            'availability': availability,
            'parameters': parameters or [],
            'enabled': enabled
        }
        return func
    
    return decorator